### PR TITLE
Fix Optimize in Anomaly Task

### PR DESCRIPTION
**Pull request by ashwinvaidya17**
### Summary

- Use `val_dataloader` for NNCF optimization
- Fixes: CVS-141031

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/anomaly/openvino_model.py [hunk 1]
```diff
@@ -133,6 +137,55 @@ def _create_model(self) -> Model:
...
+            nodes = model.get_ops()
+            return any(op.get_type_name() == "FakeQuantize" for op in nodes)
+
+        ov_model = openvino.Core().read_model(self.model_name)
+
+        if check_if_quantized(ov_model):
+            msg = "Model is already optimized by PTQ"
+            raise RuntimeError(msg)
+
+        train_dataset = data_module.val_dataloader()
```
Comment (ashwinvaidya17) [inline src/otx/algo/anomaly/openvino_model.py]: Since this is the only line that is different is there a better way to update this method without duplicating the code?

Comment (harimkang) [reply src/otx/algo/anomaly/openvino_model.py]: Is there any other reason why Anomaly specifically uses `val_dataloader` in anomaly?
@sovrasov What do you think about this? If this is what we need, shouldn't we be able to specify a subset in `OVModel.optimize`?

Comment (harimkang) [inline src/otx/algo/anomaly/openvino_model.py]:
```suggestion
        val_dataset = data_module.val_dataloader()
```

File: src/otx/algo/anomaly/openvino_model.py [hunk 2]
```diff
@@ -133,6 +137,55 @@ def _create_model(self) -> Model:
...
+
+        ptq_config_from_ir = self._read_ptq_config_from_ir(ov_model)
+        if ptq_config is not None:
+            ptq_config_from_ir.update(ptq_config)
+            ptq_config = ptq_config_from_ir
+        else:
+            ptq_config = ptq_config_from_ir
+
+        quantization_dataset = nncf.Dataset(train_dataset, self.transform_fn)  # type: ignore[attr-defined]
```
Comment (harimkang) [inline src/otx/algo/anomaly/openvino_model.py]:
```suggestion
        quantization_dataset = nncf.Dataset(val_dataset, self.transform_fn)  # type: ignore[attr-defined]
```

**Review by harimkang — DISMISSED** [review]

**Comment by chuneuny-emily** [comment]
@ashwinvaidya17 - Are we using test set for optimization? Did you check optimization works fine with this fix?

File: src/otx/algo/anomaly/openvino_model.py (hunk 1 above)
Comment (ashwinvaidya17) [reply src/otx/algo/anomaly/openvino_model.py]: I am not the expert here. I just referred to Adrian's code here https://github.com/openvinotoolkit/anomalib/blob/e5b91d6a32c469ec0b9c4c161272b68114ba6fa3/src/anomalib/models/components/base/export_mixin.py#L245
The only difference is that the val dataset also contains the anomaly class.

**Comment by ashwinvaidya17** [comment]
> @ashwinvaidya17 - Are we using test set for optimization? Did you check optimization works fine with this fix?

I tried optimization on one category (`hazelnut_large`) it seems to be working fine. I think it does use the test set for optimization as the val split is set to test set. Ideally it would be better if we could split the test set into val set and test set.

**Review by harimkang — APPROVED** [review]

**Review by chuneuny-emily — APPROVED** [review]
